tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/base/NaorPinkasOt.java

```diff
@@ -0,0 +1,323 @@
+package dk.alexandra.fresco.tools.ot.base;
+
+import dk.alexandra.fresco.framework.MaliciousException;
+import dk.alexandra.fresco.framework.network.Network;
+import dk.alexandra.fresco.framework.util.AesCtrDrbg;
+import dk.alexandra.fresco.framework.util.ByteArrayHelper;
+import dk.alexandra.fresco.framework.util.Drbg;
+import dk.alexandra.fresco.framework.util.Drng;
+import dk.alexandra.fresco.framework.util.DrngImpl;
+import dk.alexandra.fresco.framework.util.ExceptionConverter;
+import dk.alexandra.fresco.framework.util.Pair;
+import dk.alexandra.fresco.framework.util.StrictBitVector;
+import dk.alexandra.fresco.tools.cointossing.CoinTossing;
+import java.math.BigInteger;
+import java.security.AlgorithmParameterGenerator;
+import java.security.AlgorithmParameters;
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.util.Arrays;
+import javax.crypto.spec.DHParameterSpec;
+
+/**
+ * Implementation of the Naor-Pinkas OT.
+ */
+public class NaorPinkasOt implements Ot {
+  private final int myId;
+  private final int otherId;
+  private final Network network;
+  private final Drng randNum;
+  private final Drbg randBit;
+  private final MessageDigest hashDigest;
+  private static final String HASH_ALGORITHM = "SHA-256";
+  /**
+   * The PRG algorithm used internally by Java to generate the Diffie-Hellman
+   * parameters based on some seed. This PRG MUST only be used for this purpose
+   * since SHA1 is insecure in the general case.
+   */
+  private static final String PRG_ALGORITHM = "SHA1PRNG";
+  private static final int DIFFIE_HELLMAN_SIZE = 2048;
+  private DHParameterSpec params;
+
+  /**
+   * Constructs a Naor-Pinkas OT instance by interactively and securely sampling
+   * the underlying Diffie-Hellman parameters.
+   *
+   * @param myId
+   *          The ID of the calling party
+   * @param otherId
+   *          The ID of the other party
+   * @param rand
+   *          The calling party's secure randomness generator
+   * @param network
+   *          The underlying network to use
+   */
+  public NaorPinkasOt(int myId, int otherId, Drbg rand, Network network) {
+    this(myId, otherId, rand, network, null);
+    setDhParams(computeSecureDhParams());
```

Comment:
There is quite a big gap depending on the DhParams is being computed or not. I would argue that this constructor should be removed.

This one is only used in the mascot demo - which might even not make reasonable sense to keep

Reply:
I disagree. It depends on the security requirements of the user whether the DH parameters should be computed on the fly, or just use the static ones we have supplied. Until we actually agree that it is "secure enough" to force the user to use the parameters we have generated this constructor should remain. This, however, might be sensibly argued since the parameters are computed using build-in Java generation based on a static seed.

Reply:
My point was in the context of the constructor. There are two very dissimilar constructors. And the class really does not allow null in the DhParams.

They are not really similar. My suggestion would be to move the construction of the DhParams elsewhere.

Reply:
This has been fixed in 3a44eb50. 
I have made a separate class for constructing DH parameters and left a constructor for Naor-Pinkas OT that requires parameters to be given. 
Notice that the coverage of this new class is low. This is because the secure generation of DH parameters is very slow and so the tests do not use the secure generation. However, for security reasons it is important to be able to generate these parameters securely using coin tossing. 
A possible solution is to adjust the size of the DH group using reflection. However, this value is currently final and static (and initialised in the field) and thus I have not been able to adjust it using reflection. 
Suggestions of how to handle the testing is welcome.